includes/utils/class-amp-image-dimension-extractor.php

```diff
@@ -2,20 +2,44 @@
 
 class AMP_Image_Dimension_Extractor {
 	static $callbacks_registered = false;
+    const STATUS_FAILED_LAST_ATTEMPT = 'failed';
+    const STATUS_IMAGE_EXTRACTION_FAILED= 'failed';
 
-	static public function extract( $url ) {
+    static public function extract( $urls ) {
```

Comment:
Thanks for your detailed thoughts on this! The main reason I brought it up is that backwards compatibility in the WordPress ecosystem is often preferred over the cost of complexity (there are a lot of reasons for this and many pros/cons on both sides).

In any case, your arguments seem reasonable and it makes sense to remove the existing filter and introduce the new batch one.

> If you agree, I can update wpcom-helper.php wpcom_amp_extract_image_dimensions_add_custom_callbacks() to use the batch filter instead of the original one.

Yes, it'd be great to see how the existing callbacks we have look with the batch approach instead.